lib/mode-reset/index.js

```diff
@@ -221,34 +222,47 @@ ResetPanel.prototype.valid = function () {
   widget._focusError();
 
   if (email_empty) {
+    if (validate_username) {
+      widget.emit('reset error', new ValidationError('username empty'));
+    } else {
+      widget.emit('reset error', new ValidationError('email empty'));
+    }
     widget._focusError(email_input);
     ok = false;
   }
 
   if (!email_parsed && !email_empty) {
-    if(validate_username && !username_parsed) {
-      ok = false || (validate_username && username_parsed);
-      if(!ok) widget._focusError(email_input, widget.options.i18n.t('invalid'));
+    ok = false || (validate_username && username_parsed);
+
+    if (validate_username && !username_parsed) {
+      widget.emit('reset error', new ValidationError('username invalid'));
+      widget._focusError(email_input, widget.options.i18n.t('invalid'));
+    } else if (!ok) {
+      widget.emit('reset error', new ValidationError('email invalid'));
+      widget._focusError(email_input, widget.options.i18n.t('invalid'));
     }
   }
 
   if (password_empty) {
+    widget.emit('reset error', new ValidationError('password empty'));
     widget._focusError(password_input);
     ok = false;
   }
 
   if (repeat_password_empty) {
+    widget.emit('reset error', new ValidationError('repeat password empty'));
     widget._focusError(repeat_password_input);
     ok = false;
   }
 
   if (repeat_password_input.val() !== password_input.val()) {
+    widget.emit('reset error', new ValidationError('password missmatch'));
     widget._focusError(repeat_password_input, widget.options.i18n.t('mustMatch'));
     ok = false;
   }
 
   return ok;
-}
+};
```

Comment:
maybe we could save the `error = new ValidationError('...');` and trigger the emit at the end right before the `return ok`.